docs/RFCS/20220818_tenant_version_upgrades.md

```diff
@@ -0,0 +1,475 @@
+- add the `CREATELOGIN` option to roles that already have the `CREATEROLE` role option.
+
+## How these mechanisms have been extended with multi-tenancy
+
+### Startup migrations
+
+The simpler, legacy startup migration mechanism has a trivial extension to multi-tenancy:
+
+*Every time a SQL server for a secondary tenant starts, it attempts to run whichever
+startup migrations it can for its tenant system tables.*
+
+This is checked/done over and over again every time a tenant SQL server starts.
+
+It's also possible for different tenants to "step through" their
+startup migrations at different rates, and that's OK.
+
+### New concept: separate logical versions for tenants and the KV layer
+
+In a multi-tenant deployment, it's pretty important that different tenants (customers) can
+introduce SQL features "at their own leisure".
+
+So maybe tenant A wants to consume our features at logical version X
+and tenant B at logical version Y, with X and Y far from each other.
+
+To make this possible, we introduce the following concept:
+
+*Each tenant has its own logical cluster version, separate from other tenants.*
+
+With this introduction, we now have 4 "version" values of interest:
+
+- the logical version in the system tenant, which is also
+  the logical version used / stored in KV stores.
+
+  We will call this the *storage logical version* (SLV) henceforth.
+
+- the executable version(s) used in KV nodes.
+
+  We will call this the *storage binary version* (SBV) henceforth.
+
+- the logical version in each secondary tenant (there may be several,
+  and they can also be separate from the logical version in the system
+  tenant).
+
+  We will call this the (set of) *tenant logical versions* (TLV)
+  henceforth.
+
+- the executable version used for SQL servers (which can be different
+  from that used for KV nodes,
+  [#84700](https://github.com/cockroachdb/cockroach/pull/84700)
+  notwithstanding).
+
+  We will call this the *tenant binary version* (TBV) henceforth.
+
+### Invariants in a multi-tenancy world
+
+The original single-tenancy invariants extend as follows:
+
+- on the storage cluster:
+
+  A) SLV must be permissible for current SBV
+
+     (i.e. KV nodes running at 20.2 must have SLV at 20.1-XX or 20.2-XX)
+
+  B) storage SQL+KV layers must observe max 1 version difference for SLV, and SLV must evolve monotonically.
+
+- on each tenant:
+
+  C) TLV must be permissible for current TBV
+
+     (i.e. SQL servers running at 20.2 must have TLV at 20.1-XX or 20.2-XX)
+
+  D) all SQL servers for 1 tenant must observe max 1 version difference for this tenant's TLV, and TLV must evolve monotonically.
+
+In addition to the natural extensions above, we introduce the following *new* invariant:
+
+E) the storage SLV must be greater or equal to all the tenants TLVs.
+   (conversely: we must not allow the TLVs to move past the SLV)
+
+   We need this because we want to simplify our engineering process: we would
```

Comment:
While this is correct, I generally frame this on SQL behavior vs KV API behavior rather than tenant/storage cluster:  basically, if some piece of SQL code -- running in the system tenant or guest tenant -- observes cluster verison X, that is sufficient to know that a) all other running SQL code is and will remain compatible with X and b) all of the KV API is and will remain compatible with X.